cpp/mrc/src/internal/service.cpp

```diff
@@ -17,131 +17,295 @@
+bool Service::is_service_startable() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state == ServiceState::Initialized);
+}
+
+bool Service::is_running() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return (m_state > ServiceState::Initialized && m_state < ServiceState::Completed);
+}
+
+const ServiceState& Service::state() const
+{
+    std::lock_guard<decltype(m_mutex)> lock(m_mutex);
+    return m_state;
+}
+
 void Service::service_start()
 {
-    if (forward_state(ServiceState::Running))
+    std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+    if (!this->is_service_startable())
     {
-        do_service_start();
+        throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service has already been started"));
+    }
+
+    if (forward_state(ServiceState::Starting))
+    {
+        // Unlock the mutex before calling start to avoid a deadlock
+        lock.unlock();
+
+        try
+        {
+            this->do_service_start();
+
+            // Use ensure_state here in case the service itself called stop or kill
+            this->ensure_state(ServiceState::Running);
+        } catch (...)
+        {
+            // On error, set this to completed and rethrow the error to allow for cleanup
+            this->forward_state(ServiceState::Completed);
+
+            throw;
+        }
     }
 }
 
 void Service::service_await_live()
 {
-    do_service_await_live();
+    {
+        std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+        if (this->is_service_startable())
+        {
+            throw exceptions::MrcRuntimeError(MRC_CONCAT_STR(this->debug_prefix() << " Service must be started before "
+                                                                                     "awaiting live"));
+        }
+
+        // Check if this is our first call to service_await_join
+        if (!m_service_await_live_called)
+        {
+            // Prevent reentry
+            m_service_await_live_called = true;
+
+            // We now create a promise and a future to track the completion of this function
+            Promise<void> live_promise;
+
+            m_live_future = live_promise.get_future();
+
+            // Unlock the mutex before calling await to avoid a deadlock
+            lock.unlock();
+
+            try
+            {
+                {
+                    // Now call the await join (this can throw!)
+                    this->do_service_await_live();
+                }
+
+                // Set the value only if there was not an exception
+                live_promise.set_value();
+
+            } catch (...)
+            {
+                // Join must have thrown, set the exception in the promise (it will be retrieved later)
+                live_promise.set_exception(std::current_exception());
+            }
+        }
+    }
+
+    // Wait for the future to be returned. This will rethrow and exception thrown in do_service_await_join
```

Comment:
```suggestion
    // Wait for the future to be returned. This will rethrow any exception thrown in do_service_await_join
```

Reply:
Added to review in https://github.com/nv-morpheus/MRC/pull/391